modules/core/include/opencv2/core/cvdef.h

```diff
@@ -575,14 +575,43 @@ Cv64suf;
 #  endif
 #endif
 
+/****************************************************************************************\
+*                       CV_NODISCARD_STD attribute (C++17)                               *
+* encourages the compiler to issue a warning if the return value is discarded            *
+\****************************************************************************************/
+#ifndef CV_NODISCARD_STD
+#  if defined(__has_cpp_attribute) && __has_cpp_attribute(nodiscard)
```

Comment:
`__has_cpp_attribute(nodiscard)` should be checked separately (through own `#if` statement)

Reply:
I disagree. The preprocessor follows strict C logic short-circuiting. Therefore if it doesnt support the _has_cpp_attribute quasi-macro, then the part after the && doesn't run.
In addition it is important that both are true. Otherwise I have to write duplicated code so that I correct handle the case of a compiler supporting _has_cpp_attribute but does not support the nodiscard attribute...in that case, it needs to continue on to the other `#elif` clauses and *not* exit the entire if chain.

What is your thinking to separate them?

Reply:
There are several failed build configurations on CI.

[Enjoy](https://godbolt.org/z/GnjK8sz8r).

Reply:
Thanks for repro. I can fix that. LLVM had a similar problem https://reviews.llvm.org/D57848
These outdated ​compilers with OSs past end-of-life are troublesome. ✂☠⏲

I'll check the CI results when the CI system is out of maintenance. I checked the PR on the 5 OSs that I listed. Ubuntu 14 w/ gcc 4.x is two years beyond end-of-life so I did not check with it.

Reply:
Build logs from CI should be available now.